src/simulation.jl

```diff
@@ -190,93 +252,143 @@ function compile_env(d::AbstractNamedDecapode, dec_matrices::Vector{Symbol}, con
     push!(defined_ops, op)
   end
 
-  for op in d[:op1]
-    if op == DerivOp
+  # Add in user-defined operations
+  for op in vcat(d[:op1], d[:op2])
+    if op == DerivOp || op in defined_ops || op in ARITHMETIC_OPS
       continue
     end
-
-    if(op in con_dec_operators || op in defined_ops || op in assumed_ops)
-      continue
-    end
-
     ops = QuoteNode(op)
     def = :($op = operators(mesh, $ops))
-
     push!(defs.args, def)
 
     push!(defined_ops, op)
   end
-  for op in d[:op2]
-    if op in assumed_ops || op in defined_ops
-      continue
-    end
-    ops = QuoteNode(op)
-    def = :($op = operators(mesh, $ops))
-    push!(defs.args, def)
 
-    push!(defined_ops, op)
-  end
   return defs
 end
 
-function compile_var(alloc_vectors::Vector{AllocVecCall})
-  return quote $(Expr.(alloc_vectors)...) end
+struct AmbiguousNameException <: Exception
+  name::Symbol
+  indices::Vector{Int}
+end
+
+Base.showerror(io::IO, e::AmbiguousNameException) = print(io, "Name \"$(e.name)\" is repeated at indices $(e.indices) and is ambiguous")
+
+struct InvalidDecaTypeException <: Exception
+  name::Symbol
+  type::Symbol
 end
 
-# This is the block of parameter setting inside f
-# TODO: Pass this an extra type parameter that sets the size of the Floats
-get_vars_code(d::AbstractNamedDecapode, vars::Vector{Symbol}) = get_vars_code(d, vars, Float64)
-
-function get_vars_code(d::AbstractNamedDecapode, vars::Vector{Symbol}, ::Type{stateeltype}) where stateeltype
-  stmts = map(vars) do s
-    ssymbl = QuoteNode(s)
-    if all(d[incident(d, s, :name) , :type] .== :Constant)
-    #if only(d[incident(d, s, :name) , :type]) == :Constant
-      :($s = p.$s)
-    elseif all(d[incident(d, s, :name) , :type] .== :Parameter)
-      :($s = (p.$s)(t))
-    elseif all(d[incident(d, s, :name) , :type] .== :Literal)
-      # Literals don't need assignments, because they are literals, but we stored them as Symbols.
-      # #TODO: we should fix that upstream so that we don't need this.
-      :($s = $(parse(stateeltype, String(s))))
-    else
-      # TODO: If names are not unique, then the type is assumed to be a
-      # form for all of the vars sharing a same name.
-      :($s = u.$s)
+Base.showerror(io::IO, e::InvalidDecaTypeException) = print(io, "Variable \"$(e.name)\" has invalid type \"$(e.type)\"")
+
+"""
+    get_vars_code(d::SummationDecapode, vars::Vector{Symbol}, ::Type{stateeltype}) where stateeltype
+
+This initalizes all input variables according to their Decapodes type.
+"""
+function get_vars_code(d::SummationDecapode, vars::Vector{Symbol}, ::Type{stateeltype}, code_target::GenerationTarget) where stateeltype
+  stmts = quote end
+
+  map(vars) do s
```

Comment:
Since we no longer need the result of this `map`, you can change it to `foreach`